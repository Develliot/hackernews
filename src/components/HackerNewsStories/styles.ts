import styled from 'styled-components';

export const StoryGrid = styled.div`
    display: grid;
    /* grid-template-columns: repeat(6, 1fr); */
    grid-gap: 1rem;
    justify-content: center;
    align-content: start;
    grid-auto-flow: dense;
    grid-template-columns: repeat(auto-fill, minmax(180px, 1fr));
    grid-auto-rows: minmax(180px, auto);
`;
