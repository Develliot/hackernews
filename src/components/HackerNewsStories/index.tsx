import React, { useState, useRef, FunctionComponent } from 'react';

import { useScrollPosition } from 'src/hooks/UseScrollPosition';
import HackerNewsStoryContainer from 'src/containers/HackerNewsStoryContainer';
import { ErrorMessage } from 'src/components/ErrorMessage';
import { StoryGrid } from './styles';

const MAX_STORIES = 500;
const STORIES_PER_PAGE = 24;

type Props = {
    isError: boolean;
    isLoading: boolean;
    storyIds: number[];
    retry: () => void;
};

export const HackerNewsStories: FunctionComponent<Props> = ({
    isError,
    isLoading,
    storyIds,
    retry,
}) => {
    const scrollRef = useRef<HTMLDivElement>(null);
    const [storyCount, setStoryCount] = useState(STORIES_PER_PAGE);
    const isBrowser = typeof window !== `undefined`;

    useScrollPosition(
        ({
            currPos,
        }: {
            currPos: { x: number; y: number; height: number; width: number };
        }) => {
            if (
                !isBrowser ||
                !scrollRef ||
                !scrollRef.current ||
                storyCount >= MAX_STORIES
            ) {
                return null;
            }

            const loadMore = currPos.y + currPos.height <= window.innerHeight;

            if (loadMore) {
                const calculatedCount = storyCount + STORIES_PER_PAGE;
                const limitedCount =
                    calculatedCount > MAX_STORIES
                        ? MAX_STORIES
                        : calculatedCount;
                setStoryCount(limitedCount);
            }
        },
        [storyCount],
        scrollRef,
        false,
        300
    );

    return (
        <>
            {isError ? (
                <ErrorMessage
                    errorMessage='Unable to retrieve stories'
                    retry={retry}
                />
            ) : isLoading ? (
                <StoryGrid>
                    <span>Loading stories</span>
                    {/* TODO: loading state */}
                    {/* {[...Array(24)].map((e, i) => (
                        <StoryCardLoading />
                    ))} */}
                </StoryGrid>
            ) : (
                <StoryGrid ref={scrollRef}>
                    {storyIds.slice(0, storyCount).map((storyId: number) => (
                        <HackerNewsStoryContainer
                            key={storyId}
                            storyId={storyId}
                        />
                    ))}
                </StoryGrid>
            )}
        </>
    );
};

export default HackerNewsStories;
