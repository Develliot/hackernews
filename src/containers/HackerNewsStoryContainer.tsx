import React, { useEffect, FunctionComponent, memo } from 'react';
import { useGetRequest } from 'src/hooks/useGetRequest';
import { HackerNewsStoryCard } from 'src/components/HackerNewsStoryCard';

const HACKER_NEWS_ITEM_URL = 'https://hacker-news.firebaseio.com/v0/item/';

export type StoryType = {
    by: string;
    descendants: number;
    id: number;
    kids: number[];
    score: number;
    time: number;
    title: string;
    type: string;
    url: string;
};

type Props = {
    storyId: number;
};

export const HackerNewsStoryContainer: FunctionComponent<Props> = ({
    storyId,
}) => {
    const url = `${HACKER_NEWS_ITEM_URL}${storyId}.json`;
    const [story, isLoading, isError, setUrl] = useGetRequest(url, []);

    // run once
    useEffect(() => {
        setUrl(url);
    }, [setUrl]);

    const retry = (): void => {
        setUrl(url);
    };

    return (
        <HackerNewsStoryCard
            isError={isError}
            isLoading={isLoading}
            retry={retry}
            story={story}
        />
    );
};

export default memo(HackerNewsStoryContainer);
