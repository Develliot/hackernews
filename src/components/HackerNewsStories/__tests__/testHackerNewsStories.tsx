import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import HackerNewsStories, {
    STORIES_PER_PAGE,
    checkForScrollToBottom,
} from '../';
import HackerNewsStoryContainer from 'src/containers/HackerNewsStoryContainer';
import UseScrollPosition from 'src/hooks/UseScrollPosition';

jest.mock('src/hooks/UseScrollPosition');

describe('HackerNewsStoryContainer', () => {
    const mockProps = {
        isError: false,
        isLoading: false,
        storyIds: Array.from(Array(500).keys()),
        retry: () => {},
    };

    beforeEach(() => {
        (UseScrollPosition as jest.Mock<any>).mockClear();
    });

    it('renders correctly when loading', () => {
        const wrapper: ShallowWrapper = shallow(
            <HackerNewsStories {...mockProps} isLoading={true} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with error', () => {
        const wrapper: ShallowWrapper = shallow(
            <HackerNewsStories {...mockProps} isError={true} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with children', () => {
        const wrapper: ShallowWrapper = shallow(
            <HackerNewsStories {...mockProps} />
        );

        act(() => {
            expect(wrapper.find(HackerNewsStoryContainer)).toHaveLength(
                STORIES_PER_PAGE
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('checks if scroll to bottom correctly', () => {
        expect(checkForScrollToBottom(0, 1000, 100)).toEqual(false);
        expect(checkForScrollToBottom(-500, 1000, 100)).toEqual(false);
        expect(checkForScrollToBottom(-1000, 1000, 100)).toEqual(true);
    });
});
