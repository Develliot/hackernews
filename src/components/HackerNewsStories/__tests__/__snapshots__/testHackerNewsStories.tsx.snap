// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackerNewsStoryContainer renders correctly when loading 1`] = `
<Fragment>
  <styled.div>
    <span>
      Loading stories
    </span>
  </styled.div>
</Fragment>
`;

exports[`HackerNewsStoryContainer renders correctly with children 1`] = `
<Fragment>
  <styled.div>
    <Memo(HackerNewsStoryContainer)
      key="0"
      storyId={0}
    />
    <Memo(HackerNewsStoryContainer)
      key="1"
      storyId={1}
    />
    <Memo(HackerNewsStoryContainer)
      key="2"
      storyId={2}
    />
    <Memo(HackerNewsStoryContainer)
      key="3"
      storyId={3}
    />
    <Memo(HackerNewsStoryContainer)
      key="4"
      storyId={4}
    />
    <Memo(HackerNewsStoryContainer)
      key="5"
      storyId={5}
    />
    <Memo(HackerNewsStoryContainer)
      key="6"
      storyId={6}
    />
    <Memo(HackerNewsStoryContainer)
      key="7"
      storyId={7}
    />
    <Memo(HackerNewsStoryContainer)
      key="8"
      storyId={8}
    />
    <Memo(HackerNewsStoryContainer)
      key="9"
      storyId={9}
    />
    <Memo(HackerNewsStoryContainer)
      key="10"
      storyId={10}
    />
    <Memo(HackerNewsStoryContainer)
      key="11"
      storyId={11}
    />
    <Memo(HackerNewsStoryContainer)
      key="12"
      storyId={12}
    />
    <Memo(HackerNewsStoryContainer)
      key="13"
      storyId={13}
    />
    <Memo(HackerNewsStoryContainer)
      key="14"
      storyId={14}
    />
    <Memo(HackerNewsStoryContainer)
      key="15"
      storyId={15}
    />
    <Memo(HackerNewsStoryContainer)
      key="16"
      storyId={16}
    />
    <Memo(HackerNewsStoryContainer)
      key="17"
      storyId={17}
    />
    <Memo(HackerNewsStoryContainer)
      key="18"
      storyId={18}
    />
    <Memo(HackerNewsStoryContainer)
      key="19"
      storyId={19}
    />
    <Memo(HackerNewsStoryContainer)
      key="20"
      storyId={20}
    />
    <Memo(HackerNewsStoryContainer)
      key="21"
      storyId={21}
    />
    <Memo(HackerNewsStoryContainer)
      key="22"
      storyId={22}
    />
    <Memo(HackerNewsStoryContainer)
      key="23"
      storyId={23}
    />
  </styled.div>
</Fragment>
`;

exports[`HackerNewsStoryContainer renders correctly with error 1`] = `
<Fragment>
  <ErrorMessage
    errorMessage="Unable to retrieve stories"
    retry={[Function]}
  />
</Fragment>
`;
